File: examples/video_sub.rs
//! Subscriber: opens a Wayland window and draws frames from `video_pub`.
//!
//! Run:
//!
//! ```bash
//! cargo run --release --example video_sub -- <did:key:z…>
//! ```
//!
//! Same command works whether the publisher is on the same host
//! (local SHM) or another machine (iroh). peerbus picks.

use std::time::{Duration, Instant};

use datapod::{Encoding, Grid};
use minifb::{Key, Window, WindowOptions};
use peerbus::{DatapodMsg, LocalConfig, Node};

const TOPIC: &str = "demo/video";

fn main() -> peerbus::Result<()> {
    init_tracing();
    ensure_wayland_session();

    let did = std::env::args()
        .nth(1)
        .expect("usage: video_sub <did:key:z…>");

    // Big enough for 4K RGBA frames.
    let local_cfg = LocalConfig {
        max_payload_bytes: 64 * 1024 * 1024,
        subscriber_buffer: 4,
        ..LocalConfig::default()
    };
    let node = Node::builder().local_config(local_cfg).bind()?;

    println!("subscribing to {did} on '{TOPIC}'  (Esc to quit)");
    let mut sub = node.subscriber::<DatapodMsg>(did.as_str(), TOPIC)?;

    println!("waiting for first frame …");
    let (width, height, first_frame) = loop {
        if let Some(s) = sub.take()? {
            if let Some(grid) = decode_grid_message(s.header(), s.payload())? {
                break (grid.cols as usize, grid.rows as usize, grid.data);
            }
        }
        std::thread::sleep(Duration::from_millis(20));
    };
    println!("first frame {width}x{height}");

    let mut window = Window::new(
        "peerbus video sub  (Esc to quit)",
        width,
        height,
        WindowOptions {
            resize: true,
            ..WindowOptions::default()
        },
    )
    .expect("open window");
    window.set_target_fps(60);

    let mut display_buf = vec![0u32; width * height];
    copy_payload_into(&first_frame, &mut display_buf);
    let mut bench_start = Instant::now();
    let mut bench_frames: u64 = 0;

    while window.is_open() && !window.is_key_down(Key::Escape) {
        let mut got = false;
        while let Some(s) = sub.take()? {
            if let Some(grid) = decode_grid_message(s.header(), s.payload())?
                && grid.cols as usize == width
                && grid.rows as usize == height
            {
                copy_payload_into(&grid.data, &mut display_buf);
                bench_frames += 1;
                got = true;
            }
        }
        if !got {
            std::thread::sleep(Duration::from_millis(2));
        }

        window
            .update_with_buffer(&display_buf, width, height)
            .expect("update");

        if bench_start.elapsed() >= Duration::from_secs(2) {
            let elapsed = bench_start.elapsed().as_secs_f64();
            let fps = bench_frames as f64 / elapsed;
            println!("[sub] {fps:>6.1} fps  |  datapod.Grid RGBA8");
            bench_start = Instant::now();
            bench_frames = 0;
        }
    }
    Ok(())
}

fn ensure_wayland_session() {
    if std::env::var_os("WAYLAND_DISPLAY").is_none() {
        eprintln!(
            "error: video_sub is built Wayland-only; WAYLAND_DISPLAY is not set \
             (start it inside a Wayland session, not Xorg/XWayland)"
        );
        std::process::exit(2);
    }
}

fn init_tracing() {
    // RUST_LOG=iroh=info,peerbus=debug … cargo run --example …
    use tracing_subscriber::{EnvFilter, fmt};
    let filter = EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("warn"));
    let _ = fmt().with_env_filter(filter).try_init();
}

fn decode_grid_message(
    msg: &peerbus::datapod_msg::DatapodMsgHeader,
    wire: &[u8],
) -> peerbus::Result<Option<Grid>> {
    let msg = DatapodMsg::new(msg.type_hash, wire.to_vec());
    let grid = match msg.to_datapod::<Grid>() {
        Ok(grid) => grid,
        Err(datapod::WireError::WrongTypeHash { .. }) => return Ok(None),
        Err(e) => {
            return Err(peerbus::Error::invalid_argument(format!(
                "invalid datapod.Grid wire message: {e}"
            )));
        }
    };
    if grid.encoding != Encoding::Rgba8 {
        return Ok(None);
    }
    let expected = (grid.rows as usize)
        .checked_mul(grid.cols as usize)
        .and_then(|n| n.checked_mul(4))
        .ok_or_else(|| peerbus::Error::invalid_argument("video dimensions overflow"))?;
    if grid.data.len() != expected {
        return Ok(None);
    }
    Ok(Some(grid))
}

fn copy_payload_into(payload: &[u8], dst: &mut [u32]) {
    if let Ok(typed) = bytemuck::try_cast_slice::<u8, u32>(payload)
        && typed.len() == dst.len()
    {
        dst.copy_from_slice(typed);
        return;
    }
    for (slot, chunk) in dst.iter_mut().zip(payload.chunks_exact(4)) {
        *slot = u32::from_le_bytes(chunk.try_into().unwrap());
    }
}

File: examples/python_binding/video_sub.py
#!/usr/bin/env python3
"""Python subscriber compatible with Rust `examples/video_pub.rs`.

This is intentionally headless/no-GUI so it works without pulling in
Xorg/Qt/pygame dependencies. Use the Rust `video_sub` example when you
want a Wayland window.

Run from the repo root inside `nix develop`:

    peerbus-python-develop
    peerbus-video-sub <did:key:z...>
"""

from __future__ import annotations

import argparse
import os
import struct
import time

import datapod
import peerbus


TOPIC = "demo/video"


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("peer", nargs="?", default=os.environ.get("PEER"))
    parser.add_argument("--topic", default=os.environ.get("PEERBUS_VIDEO_TOPIC", TOPIC))
    parser.add_argument("--max-payload-bytes", type=int, default=64 * 1024 * 1024)
    args = parser.parse_args()

    if not args.peer:
        raise SystemExit("usage: video_sub.py <publisher did:key:z...>")

    node = peerbus.Node(max_payload_bytes=args.max_payload_bytes, subscriber_buffer=4)
    qos = peerbus.TopicQos.latest(max_message_bytes=args.max_payload_bytes)
    sub = node.datapod_subscriber(args.peer, args.topic, qos)
    grid_type_hash = datapod.Grid.TYPE_HASH
    grid_header_size = datapod.header_size(grid_type_hash)

    print(f"python subscriber listening to {args.peer} on '{args.topic}'")
    print("waiting for first frame ...")

    width = height = None
    bench_t0 = time.monotonic()
    bench_frames = 0
    bytes_sum = 0

    while True:
        sample = sub.take_view()
        if sample is None:
            time.sleep(0.002)
            continue

        if sample.type_hash != grid_type_hash:
            continue

        wire = sample.wire_view()
        if len(wire) < grid_header_size:
            continue

        # GridHeader starts with:
        # rows: u32, cols: u32, encoding: u32, centered: u32, resolution: f64.
        # `wire` is a memoryview backed directly by the SHM slot on local
        # transport; struct.unpack_from reads the small header without copying
        # the frame payload.
        h, w, encoding_id, _centered, _resolution = struct.unpack_from("<IIIId", wire, 0)
        pixels = wire[grid_header_size:]
        if encoding_id != 11 or len(pixels) != w * h * 4:
            continue
        if (w, h) != (width, height):
            width, height = w, h
            print(f"frame size {width}x{height}")

        bench_frames += 1
        bytes_sum += len(pixels)

        elapsed = time.monotonic() - bench_t0
        if elapsed >= 2.0:
            fps = bench_frames / elapsed
            mib_s = bytes_sum / elapsed / 1_048_576.0
            print(f"[py sub] {fps:>6.1f} fps  |  {mib_s:>7.1f} MB/s  |  datapod.Grid RGBA8")
            bench_t0 = time.monotonic()
            bench_frames = 0
            bytes_sum = 0


if __name__ == "__main__":
    main()
